### ISSUE TITLE
IndexedDB tests not all auto-deleting

### ISSUE THREAD

**Issue by brettz9** [w3c/web-platform-tests#4560]
After opening a large set of IndexedDB tests in the browser, I see these are not all deleting, leading to hundreds of megabytes of data (and apparently with this large amount of data, subsequent tests were locking up).

Although I see there are mechanisms available in `support.js` and such for clean-up, might clean-up be better enforced, and not just deleting any existing database *before* running a test (like `support-promises.js` does and as do some other files which lack `support.js`), but afterward (unless of course, there is a current failure which prevents this)?

Thanks!

**Comment by inexorabletash** [comment]
cc: @pwnall 

Adding the following to tests/helpers would do the trick:

```js
  // in the indexedDB.open() success callback:
  t.add_cleanup(() => {
    db.close();
    indexedDB.deleteDatabase(db.name);
  });
```

That'd be a swell PR if anyone wants to take a stab at it.

**Comment by inexorabletash** [comment]
FYI, I'm slowly working through a PR to do this. I think I'm about halfway done - either fixing up the places that call open() directly or changing them to use one of the helper functions.

The time sinks are the files that created one database connection re-used across multiple async_test() calls. Those need to be refactored. Not rocket science, but a slog.

**Comment by inexorabletash** [comment]
Should be fixed by 70a2c07dff8604f4762f3491374a21893500ee94

**Comment by brettz9** [comment]
This is really great, thanks!!

A few files that don't delete, however:
- fire-upgradeneeded-event-exception.html
- idbfactory-open-error-properties.html
- idbfactory_open9.htm

I'm checking whether there are others, but the above at least are confirmed.

**Comment by inexorabletash** [comment]
Thanks - may have slipped through my testing regime which was sadly manual.

**Comment by inexorabletash** [comment]
Huh... yeah, those are all really obvious via code inspection. I must have just missed them. Easy fixes, though.

**Comment by brettz9** [comment]
Also note the comment I added for `idbrequest-onupgradeneeded.htm` at https://github.com/w3c/web-platform-tests/commit/70a2c07dff8604f4762f3491374a21893500ee94#diff-ae5dd9c5b5818cbaef57035385f779c6 that there should be a check for `open_request.result` as it might not be defined when an `error_func` is given.

**Comment by inexorabletash** [comment]
Reviewing the tests:

* For these three: an abort() during upgradeneeded should cancel the creation. there should not be a database left behind by this test in compliant implementations:
  * fire-upgradeneeded-event-exception.html
  * idbfactory-open-error-properties.html
  * idbfactory_open9.htm
  * ... so no changes needed there?
* For these (which use `createdb` helper) the cleanup is configured by is blocked by `add_completion_callback()` but gets blocked on open connections(s) until GC or the page closes. This can be improved by closing the connection(s).
  * idbdatabase_close.htm
  * idbfactory_open10.htm
  * idbfactory_open11.htm 
  * idbfactory_open3.htm
  * idbfactory_open6.htm
  * idbobjectstore_deleteIndex.htm
* ... and you diagnosed the other two issues.

**Comment by inexorabletash** [comment]
PR up at https://github.com/w3c/web-platform-tests/pull/5676 - @brettz9 can you review?

**Comment by brettz9** [comment]
Beautiful...working like a charm... (And yes, sorry, I must have forgotten to report back that I had since identified (and fixed) the aborted upgradeneeded's not deleting in our implementation, so those 3 tests were indeed not problematic as you say.)

Thanks so much!

(FWIW, as we are failing or timing out within a portion of the following tests (in Node), I can't be sure there weren't deleting problems there.)

'bindings-inject-key.html', 
'event-dispatch-active-flag.html',
'idb-binary-key-detached.htm',
'idbfactory-deleteDatabase-opaque-origin.html',
'idbfactory-open-opaque-origin.html',
'interfaces.html' (only one test not passing)
'interleaved-cursors.html',
'keypath-exceptions.htm',
'transaction-deactivation-timing.html',
'upgrade-transaction-deactivation-timing.html'

We also have 7 other tests (on exception order) which are technically also failing (due to our lack of precise control of transaction timeouts), but fudging `setTimeout` gets them to pass, and seeing no deleting problems there.

But all other tests are now passing and they have no problems with delete.

**Comment by brettz9** [comment]
Would you be able to push your fixes to `master`? All look good...

**Comment by inexorabletash** [comment]
Can you give an affirmative comment on the PR itself, so we've got a record there that someone looked at them?

**Comment by brettz9** [comment]
Comments were added to the PR and as it was merged and the concerns here addressed, closing...  

There are some cases still where deletions *before* tests are not occurring, but as this is a separate issue, is one for which it is not as trivial to quickly identify failing files, and as it would require refactoring older tests to use the support files and should be of lesser concern now (with complete auto-deleting coverage), assuming the tests can at least run to the clean-up stage in the implementation, I won't bother with a new issue. Thanks again!
